### PR TITLE
Add project, phase and story apis

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:20301)](https://staging.teamyapp.com/teams/107/tasks/20301)

### CODE REVIEW

File: core/api/gql/phase.go [hunk 1]
```diff
@@ -85,41 +117,196 @@ func (m Mutation) UpdatePhase(ct context.Context, args struct {
...
+			internalErr := errs.NewError(
+				errs.InvalidArgument,
+				internalErr.Error(),
+			)
+			m.deps.logger.ErrorWithContext(ct, internalErr)
+			return Phase{}, errs.ToResolverErr(internalErr)
+		}
+
+		storyIDs[index] = id
+	}
```
Comment (naturetime10) [inline core/api/gql/phase.go]: Use `collect.MapWithErr`

File: core/api/gql/phase.go [hunk 2]
```diff
@@ -85,41 +117,196 @@ func (m Mutation) UpdatePhase(ct context.Context, args struct {
...
```
Comment (naturetime10) [inline core/api/gql/phase.go]: Use `collect.MapWithErr`

File: core/api/gql/story.go [hunk 3]
```diff
@@ -81,41 +123,202 @@ func (m Mutation) UpdateStory(ct context.Context, args struct {
...
+		if internalErr != nil {
+			internalErr := errs.NewError(
+				errs.InvalidArgument,
+				internalErr.Error(),
+			)
+			m.deps.logger.ErrorWithContext(ct, internalErr)
+			return Story{}, errs.ToResolverErr(internalErr)
+		}
+		taskIDs[index] = id
+	}
```
Comment (naturetime10) [inline core/api/gql/story.go]: collect.MapWithErr

File: core/api/gql/story.go [hunk 4]
```diff
@@ -81,41 +123,202 @@ func (m Mutation) UpdateStory(ct context.Context, args struct {
...
+
+	taskID, internalErr := fromGraphQLID(args.TaskID)
+	if internalErr != nil {
+		internalErr := errs.NewError(
+			errs.InvalidArgument,
+			internalErr.Error(),
+		)
+		m.deps.logger.ErrorWithContext(ct, internalErr)
+		return Story{}, errs.ToResolverErr(internalErr)
+	}
```
Comment (naturetime10) [inline core/api/gql/story.go:279]: collect.MapWithErr

File: core/api/gql/story.go [hunk 5]
```diff
@@ -81,41 +123,202 @@ func (m Mutation) UpdateStory(ct context.Context, args struct {
...
+		if internalErr != nil {
+			internalErr := errs.NewError(
+				errs.InvalidArgument,
+				internalErr.Error(),
+			)
+			m.deps.logger.ErrorWithContext(ct, internalErr)
+			return Story{}, errs.ToResolverErr(internalErr)
+		}
+		taskIDs[index] = id
+	}
```
Comment (naturetime10) [inline core/api/gql/story.go]: collect.MapWithErr

File: core/dao/sqldb/project_phase_relation.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+func (p *ProjectPhaseRelation) CreateProjectPhaseRelation(ct context.Context, tx *transaction.Transaction, projectPhaseRelation entity.ProjectPhaseRelation) *errs.Error {
+	panic("unimplemented")
+}
+
+func (p *ProjectPhaseRelation) DeleteProjectPhaseRelation(ct context.Context, tx *transaction.Transaction, projectID uint64, phaseID uint64) *errs.Error {
+	panic("unimplemented")
+}
+
+var _ dao.ProjectPhaseRelation = (*ProjectPhaseRelation)(nil)
```
Comment (naturetime10) [inline core/dao/sqldb/project_phase_relation.go:28]:
```suggestion
var _ dao.ProjectPhaseRelation = (*ProjectPhaseRelation)(nil)

func (p *ProjectPhaseRelation) FindPhaseIDsByProjectIDWithTx(ct context.Context, tx *transaction.Transaction, projectID uint64) ([]uint64, *errs.Error) {
	panic("unimplemented")
}

func (p *ProjectPhaseRelation) CreateProjectPhaseRelation(ct context.Context, tx *transaction.Transaction, projectPhaseRelation entity.ProjectPhaseRelation) *errs.Error {
	panic("unimplemented")
}

func (p *ProjectPhaseRelation) DeleteProjectPhaseRelation(ct context.Context, tx *transaction.Transaction, projectID uint64, phaseID uint64) *errs.Error {
	panic("unimplemented")
}
```

File: core/dao/sqldb/project_story_relation.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+func (p *ProjectStoryRelation) CreateProjectStoryRelation(ct context.Context, tx *transaction.Transaction, projectStoryRelation entity.ProjectStoryRelation) *errs.Error {
+	panic("unimplemented")
+}
+
+func (p *ProjectStoryRelation) DeleteProjectStoryRelation(ct context.Context, tx *transaction.Transaction, projectID uint64, storyID uint64) *errs.Error {
+	panic("unimplemented")
+}
+
+var _ dao.ProjectStoryRelation = (*ProjectStoryRelation)(nil)
```
Comment (naturetime10) [inline core/dao/sqldb/project_story_relation.go:28]:
```suggestion
var _ dao.ProjectStoryRelation = (*ProjectStoryRelation)(nil)

type ProjectStoryRelation struct {
	transactionFactory transaction.Factory
}

func (p *ProjectStoryRelation) FindStoryIDsByProjectIDWithTx(ct context.Context, tx *transaction.Transaction, projectID uint64) ([]uint64, *errs.Error) {
	panic("unimplemented")
}

func (p *ProjectStoryRelation) CreateProjectStoryRelation(ct context.Context, tx *transaction.Transaction, projectStoryRelation entity.ProjectStoryRelation) *errs.Error {
	panic("unimplemented")
}

func (p *ProjectStoryRelation) DeleteProjectStoryRelation(ct context.Context, tx *transaction.Transaction, projectID uint64, storyID uint64) *errs.Error {
	panic("unimplemented")
}
```

File: core/service/project.go [hunk 8]
```diff
@@ -0,0 +1,769 @@
...
+}
+
+type UpdateStoryInput struct {
+	Name     string
+	OwnerID  uint64
+	Status   entity.StoryStatus
+	Priority entity.Priority
+}
+
+func (p *Project) FindStoriesByProjectID(ct context.Context, projectID uint64) ([]entity.Story, *errs.Error) {
```
Comment (naturetime10) [inline core/service/project.go:50]: I think it's better to have `ProjectService`, `StoryService`, and `PhaseService` to decouple the logic.

File: core/service/project.go [hunk 9]
```diff
@@ -0,0 +1,769 @@
...
+	OwnerID  uint64
+	Priority entity.Priority
+}
+
+type UpdateStoryInput struct {
+	Name     string
+	OwnerID  uint64
+	Status   entity.StoryStatus
+	Priority entity.Priority
+}
```
Comment (naturetime10) [inline core/service/project.go]: Move inputs before service struct.

File: core/service/project.go [hunk 10]
```diff
@@ -0,0 +1,769 @@
...
+		ExpectedStartAt: input.ExpectedStartAt,
+		ExpectedEndAt:   input.ExpectedEndAt,
+		CreatorID:       userID,
+		CreatedAt:       time.Now(),
+	}
+
+	transactionErr := txCtx.WithTransactions(false, func(tx *cloudTransaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		return p.projectDao.CreateProject(ct, tx, project)
+	})
+
```
Comment (naturetime10) [inline core/service/project.go]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/api/gql/phase.go (hunk 1 above)
Comment (duanxianbo) [reply core/api/gql/phase.go]: what is that for and what can we save?

Comment (duanxianbo) [reply core/api/gql/phase.go]: I see

File: core/dao/sqldb/project_story_relation.go [hunk 11]
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/teamyapp/teamy-backend/core/dao"
 	"github.com/teamyapp/teamy-backend/core/entity"
 )
+var _ dao.ProjectStoryRelation = (*ProjectStoryRelation)(nil)
```
Comment (naturetime10) [inline core/dao/sqldb/project_story_relation.go]:
```suggestion

var _ dao.ProjectStoryRelation = (*ProjectStoryRelation)(nil)
```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**